fuzzers/main.py

```diff
@@ -0,0 +1,949 @@
+
+    repo = Repository(REPO_DIR, REPO_URL)
+    repo.clone_if_need_be()
+
+    corpus = Corpus(CORPUS_DIR, bucket)
+
+    while True:
+        if pause_exit_spot(pause_evt, resume_evt, exit_evt):
+            return
+
+        sync_log_files = []
+        date = datetime.datetime.utcnow().strftime('%Y-%m-%d')
+
+        # Read the configuration from the repository
+        cfg = repo.latest_config('master')
+        cfg_for = {
+            b['name']: repo.latest_config(b['name']) for b in cfg['branch']
+        }
+
+        # Figure out which targets we want to run
+        branches = random_weighted(cfg['branch'], NUM_FUZZERS, "branches")
+        targets = [
+            random_weighted(cfg_for[b['name']]['target'], 1, "target")[0]
+            for b in branches
+        ]
+
+        # Synchronize the relevant corpuses
+        corpus.stop_synchronizing()
+        corpus.update()
+        for targ in set((t['crate'], t['runner']) for t in targets):
+            log_path = pathlib.Path('sync') / date / targ[0] / targ[1] / str(
+                uuid.uuid4())
+            utils.mkdirs((LOGS_DIR / log_path).parent)
+            # pylint: disable=consider-using-with
+            corpus.synchronize(targ[0], targ[1],
+                               open(LOGS_DIR / log_path, 'a', encoding='utf-8'))
+            # pylint: enable=consider-using-with
+            sync_log_files.append(log_path)
+            if pause_exit_spot(pause_evt, resume_evt, exit_evt):
+                return
+
+        # Initialize the fuzzers
+        fuzzers = []
+        for (branch, target) in zip(branches, targets):
+            worktree = repo.worktree(branch['name'])
+            log_path = pathlib.Path(
+                'fuzz') / date / target['crate'] / target['runner'] / str(
+                    uuid.uuid4())
+            log_file = LOGS_DIR / log_path
+            utils.mkdirs(log_file.parent)
+            fuzzers.append(
+                FuzzProcess(corpus_vers=corpus.version,
+                            branch=branch,
+                            target=target,
+                            repo_dir=worktree,
+                            log_relpath=log_path,
+                            log_fullpath=log_file))
+
+        # Build the fuzzers
+        for fuzzer in fuzzers:
+            fuzzer.build()
+            if pause_exit_spot(pause_evt, resume_evt, exit_evt):
+                return
+
+        # Start the fuzzers
+        atexit.register(kill_fuzzers, bucket, fuzzers)
+        for fuzzer in fuzzers:
+            fuzzer.start(corpus)
+
+        # Wait until something happens
+        started = time.monotonic()
+        last_sync_file_upload = started
+        next_restart = started + AUTO_REFRESH_INTERVAL.total_seconds()
+        while time.monotonic() < next_restart:
+            # Avoid busy-looping by sleeping 100ms between each loop run
+            time.sleep(0.1)
+
+            # Exit event happened?
+            if exit_evt.is_set():
```

Comment:
`if exit_evt.wait(timeout=1):` and you won’t need the `time.sleep` plus will exit as soon as exit is set.

(Ideally perhaps we would want to change the `Event`s with a single `Condition` which would allow us to wait on only one thing).

Reply:
Dunno if it’s more Pythonic, I also don’t have strong opinions, it’s just how I’d write it.

Reply:
Gotcha, going with a bit more comments alongside your suggestion :)